x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1592,42 +1605,82 @@ public boolean test(String profile, InetSocketAddress peerAddress) {
                             + "[xpack.security.http.ssl.key] or [xpack.security.http.ssl.keystore.path] setting"
                     );
                 }
+                if (SSLService.isSSLClientAuthEnabled(sslConfiguration)) {
+                    populateClientCertificate = (channel, threadContext) -> extractClientCertificates(logger, threadContext, channel);
+                } else {
+                    populateClientCertificate = (channel, threadContext) -> {};
+                }
             } else {
                 sslConfiguration = null;
+                populateClientCertificate = (channel, threadContext) -> {};
             }
+            final TriConsumer<BasicHttpRequest, Channel, ThreadContext> populateThreadContext = (httpRequest, channel, threadContext) -> {
+                dispatcherContext.accept(httpRequest, threadContext);
+                populateClientCertificate.accept(channel, threadContext);
+                RemoteHostHeader.process(channel, threadContext);
+            };
```

Comment:
The thread context before running the authentication in the channel handler. It reuses that "dispatch" context that's a sort of baseline (contains selected HTTP header values).